include/circt/Dialect/FIRRTL/FIRRTLExpressions.td

```diff
@@ -642,3 +642,71 @@ def BitCastOp: FIRRTLOp<"bitcast", [NoSideEffect]> {
 
   let assemblyFormat = "$input attr-dict `:` functional-type($input, $result)";
 }
+
+//===----------------------------------------------------------------------===//
+// Cross Module Reference (XMR) Operations
+//===----------------------------------------------------------------------===//
+
+class RefTypeConstraint<string ref, string base>
+  : TypesMatchWith<"reference base type should match",
+                   ref, base,
+                   "$_self.cast<RefType>().getType()">;
+
+class RefSingleUserConstraint<string ref>
+  : PredOpTrait<"reference port operand cannot be reused by any other op",
+                CPred<"$" # ref # ".hasOneUse()">>;
+
+class RefInstancePortConstraint<string ref>
+  : PredOpTrait<"it cannot have module port as an operand because it implies"
+  " upward reference XMR, which are not supported."
+  " The reference operand must be a port from firrtl.instance",
+                CPred<"!($" # ref # ".isa<BlockArgument>())">>;
+
+class RefPortConstraint<string ref>
+  : PredOpTrait<"reference port operand can only be a module or instance port",
+                CPred<"($" # ref # ".isa<BlockArgument>()"
+                " || (isa<InstanceOp>($" # ref # ".getDefiningOp())))">>;
+
+def RefResolveOp: FIRRTLOp<"ref.resolve", [RefTypeConstraint<"ref","result">,
+                                    RefSingleUserConstraint<"ref">,
+                                    RefInstancePortConstraint<"ref">]> {
```

Comment:
The single-user constraint here seems unnecessarily conservative. I am okay with starting with this constraint, though.

Reply:
So, the single user constraint ensures that a reference port can resolve to a single XMR. Otherwise if there are multiple connections to the same ref port, it might be difficult to establish the data flow between the `ref.resolve` and `ref.send/recv` . We can get rid of it if we donot want a 1-1 correspondence between the refports and the sv XMR.